File: 剑指offer/22. 链表中倒数第k个节点.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def getKthFromEnd(self, head: ListNode, k: int) -> ListNode:
        former, latter = head, head
        for i in range(k):
            former = former.next
        while former:
            former,latter = former.next,latter.next
        return latter
    
#时间复杂度：O(N)
#空间复杂度：O(1)
https://leetcode-cn.com/problems/lian-biao-zhong-dao-shu-di-kge-jie-dian-lcof/solution/mian-shi-ti-22-lian-biao-zhong-dao-shu-di-kge-j-11/

File: 剑指offer/C++_Version/22. 链表中倒数第k个节点.cpp
//方法1
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* getKthFromEnd(ListNode* head, int k)
    {
        ListNode *former = head;
        ListNode *latter = head;
        for(int i=0; i<k; i++)
        {
            former = former->next;
        }
        while(former!=nullptr)
        {
            former = former->next;
            latter = latter->next;
        }
        return latter;
    }
};
